src/core/service.c

```diff
@@ -3249,6 +3249,7 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                         return;
 
                 s->main_pid = 0;
+                s->main_pid_known = false;
```

Comment:
> Re-watching pids on cgroup v1 (needed because of unreliability of cgroup empty notifications in containers) is handled bellow at the end of service_sigchld_event() and depends on value main_pid_known flag.

Can you attempt an alternative solution that doesn't involve marking the main PID as invalid? That doesn't seem correct to me. A better way would be to simply make re-watching PIDs not depend on it.